add priority draft (#1010)
httpwg/http-extensions: draft-ietf-httpbis-priority.md

so a server can determine if it should respect the HTTP/2 scheme before
building state.

# Priority Parameters

The priority information is a sequence of key-value pairs, providing room for
future extensions. Each key-value pair represents a priority parameter.

The Priority HTTP header field is an end-to-end way to transmit this set of
parameters when a request or a response is issued. In order to reprioritize a
request, HTTP-version-specific frames are used by clients to transmit the
same information on a single hop.  If intermediaries want to specify
prioritizaton on a multiplexed HTTP connection, it SHOULD use a
PRIORITY_UPDATE frame and SHOULD NOT change the Priority header field.

In both cases, the set of priority parameters is encoded as a Structured Headers
Dictionary ({{!STRUCTURED-HEADERS}}).

This document defines the urgency(`u`) and incremental(`i`) parameters. When
used, these parameters MUST be accompanied by values. When any of the defined
parameters are omitted, or if the Priority header field is not used, their
default values SHOULD be applied.

Unknown parameters, parameters with out-of-range values or values of unexpected
types MUST be ignored.

## urgency

> The urgency(`u`) parameter takes an integer between 0 and 7, in descending order of

Reviewer: Q: why not something like

```
~~~
urgency = sh-integer
~~~
```

|               7 | background ({{background}})       |
{: #urgencies title="Urgencies"}

The value is encoded as an sh-integer. The default value is 1.

A server SHOULD transmit HTTP responses in the order of their urgency values.
The lower the value, the higher the precedence.

The following example shows a request for a CSS file with the urgency set to
`0`:

~~~ example
:method = GET
:scheme = https
:authority = example.net
:path = /style.css
priority = u=0
~~~

The definition of the urgencies and their expected use-case are described below.
Endpoints SHOULD respect the definition of the values when assigning urgencies.

### prerequisite

The prerequisite urgency (value 0) indicates that the response prevents other
responses with an urgency of prerequisite or default from being used until it
is fully transmitted.

For example, use of an external stylesheet can block a web browser from
rendering the HTML. In such case, the stylesheet is given the prerequisite
urgency.

### default

The default urgency (value 1) indicates a response that is to be used as it is
delivered to the client, but one that does not block other responses from being
used.

> For example, when a user using a web browser navigates to a new HTML document,

Reviewer: ```suggestion
For example, when a user navigates with a web browser to a new HTML document,
```

the request for that HTML is given the default urgency.  When that HTML document

> uses a custom font, the request for that custom font SHOULD also be given the

Reviewer: We're using SHOULD in an example: is that ok?

default urgency.  This is because the availability of the custom font is likely
a precondition for the user to use that portion of the HTML document, which is
to be rendered by that font.

### supplementary

The supplementary urgencies (values 2 to 6) indicate a response that is helpful
to the client using a composition of responses, even though the response itself
is not mandatory for using those responses.

For example, inline images (i.e., images being fetched and displayed as part of
the document) are visually important elements of an HTML document.  As such,
users will typically not be prevented from using the document, at least to some
degree, before any or all of these images are loaded. Display of those images
are thus considered to be an improvement for visual clients rather than a
prerequisite for all user agents.  Therefore, such images will be given the
supplementary urgency.

Values between 2 and 6 are used to represent this urgency, to provide
flexibility to the endpoints for giving some responses more or less precedence
than others that belong to the supplementary group. {{merging}} explains how
these values might be used.

> Clients SHOULD NOT use values 2 and 6.  Servers MAY use these values to

Reviewer: ```suggestion
Clients SHOULD NOT use the boundary values 2 and 6.  Servers MAY use these values to
```

prioritize a response above or below other supplementary responses.

> Clients MAY use values 3 to indicate that a request is given relatively high

Reviewer: it's not clear to me if clients MAY use `4`.

priority, or 5 to indicate relatively low priority, within the supplementary
urgency group.

For example, an image certain to be visible at the top of the page, might be
assigned a value of 3 instead of 4, as it will have a high visual impact for the
user.  Conversely, an asynchronously loaded JavaScript file might be assigned an
urgency value of 5, as it is less likely to have a visual impact.

When none of the considerations above is applicable, the value of 3 SHOULD be
used.

### background

The background urgency (value 7) is used for responses of which the delivery can
be postponed without having an impact on using other responses.

As an example, the download of a large file in a web browser would be assigned
the background urgency so it would not impact further page loads on the same
connection.

## incremental

> The incremental(`i`) parameter takes an sh-boolean as the value that indicates if

Reviewer: Q: won't we use something like that?

```
~~~
incremental = sh-boolean
~~~
```

responses that share the same urgency.

A server SHOULD transmit non-incremental responses one by one, preferably in the
order the requests were generated.  Doing so maximizes the chance of the client
making progress in using the composition of the HTTP responses at the earliest
moment.

The following example shows a request for a JPEG file with the urgency parameter
set to `4` and the incremental parameter set to `1`.

~~~ example
:method = GET
:scheme = https
:authority = example.net
:path = /image.jpg
priority = u=4, i=?1
~~~

## Defining New Parameters

When attempting to extend priorities, care must be taken to ensure any use of
existing parameters are either unchanged or modified in a way that is backwards
compatible for peers that are unaware of the extended meaning.

# The Priority HTTP Header Field

The Priority HTTP header field can appear in requests and responses. A client
uses it to specify the priority of the response. A server uses it to inform
the client that the priority was overwritten. An intermediary can use the
Priority information from client requests and server responses to correct or
amend the precedence to suit it (see {{merging}}).

> The Priority header field is an end-to-end signal of the request priority from

Reviewer: iiuc: if it's `end-to-end`, I'd use "user-agent" and "origin server" instead of client and server.

the client or the response priority from the server.

As is the ordinary case for HTTP caching ({{?RFC7234}}), a response with a
Priority header field might be cached and re-used for subsequent requests.

> When an origin server generates the Priority response header field based on

Reviewer: Q: is "Priority response header field"  different from "Priority header field"?

properties of an HTTP request it receives, the server is expected to control the
cacheability or the applicability of the cached response, by using header fields
that control the caching behavior (e.g., Cache-Control, Vary).

# Reprioritization

After a client sends a request, it may be beneficial to change the priority of

> the response. As an example, a web browser might issue a prefetch request for

Reviewer: Do we want to reference https://www.w3.org/TR/resource-hints/ or some related resource?

a JavaScript file with the urgency parameter of the Priority request header
field set to `u=7` (background). Then, when the user navigates to a page which
references the new JavaScript file, while the prefetch is in progress, the
browser would send a reprioritization frame with the priority field value
set to `u=0` (prerequisite).

In HTTP/2 and HTTP/3, after a request message is sent on a stream, the stream
transitions to a state that prevents the client from sending additional
frames on the stream. Therefore, a client cannot reprioritize a response by
using the Priority header field.  Modifying this behavior would require a
semantic change to the protocol, but this is avoided by restricting the
stream on which a PRIORITY_UPDATE frame can be sent. In HTTP/2 the frame
is on stream zero and in HTTP/3 it is sent on the control stream
({{!I-D.ietf-quic-http}}, Section 6.2.1).

This document specifies a new PRIORITY_UPDATE frame type for HTTP/2

> ({{!RFC7540}}) and HTTP/3 ({{!I-D.ietf-quic-http}}) which enables

Reviewer: ```suggestion
({{!HTTP2=RFC7540}}) and HTTP/3 ({{!I-D.ietf-quic-http}}) which enables
```

priority = u=2
~~~

the intermediary's understanding of the urgency is promoted from `4` to `2`,
because the server-provided value overrides the value provided by the client.
The incremental value continues to be `1`, the value specified by the client,
as the server did not specify the incremental(`i`) parameter.

# Security Considerations

## Fairness {#fairness}

As a general guideline, a server SHOULD NOT use priority information for making
schedule decisions across multiple connections, unless it knows that those
connections originate from the same client. Due to this, priority information
conveyed over a non-coalesced HTTP connection (e.g., HTTP/1.1) might go unused.

The remainder of this section discusses scenarios where unfairness is
problematic and presents possible mitigations, or where unfairness is desirable.

TODO: Discuss if we should add a signal that mitigates this issue. For example,
we might add a SETTINGS parameter that indicates the next hop that the
connection is NOT coalesced (see https://github.com/kazuho/draft-kazuho-httpbis-priority/issues/99).

### Coalescing Intermediaries

When an intermediary coalesces HTTP requests coming from multiple clients into

> one HTTP/2 or HTTP/3 connection going to the backend server, requests that

Reviewer: ```suggestion
one HTTP/2 or HTTP/3 connection going to the origin server, requests that
```

originate from one client might have higher precedence than those coming from
others.

> It is sometimes beneficial for the server running behind an intermediary to obey

Reviewer: ```suggestion
It is sometimes beneficial for the origin server running behind an intermediary to obey
```

to the value of the Priority header field. As an example, a resource-constrained
server might defer the transmission of software update files that would have the
background urgency being associated. However, in the worst case, the asymmetry
between the precedence declared by multiple clients might cause responses going

> to one end client to be delayed totally after those going to another.

Reviewer: ```suggestion
to one user agent to be delayed totally after those going to another.
```

> In order to mitigate this fairness problem, when a server responds to a request

Reviewer: The explanation can be improved.

that is known to have come through an intermediary, the server SHOULD prioritize
the response as if it was assigned the priority of  `u=1, i=?1`
(i.e. round-robin) regardless of the value of the Priority header field being
transmitted, unless the server knows the intermediary is not
coalescing requests from multiple clients.

A server can determine if a request came from an intermediary through

> configuration, or by consulting if that request contains one of the following

Reviewer: ```suggestion
configuration, or by checking if the request contains one of the following
```

header fields:

* CDN-Loop ({{?RFC8586}})
* Forwarded, X-Forwarded-For ({{?RFC7239}})
* Via ({{?RFC7230}}, Section 5.7.1)

Responding to requests coming through an intermediary in a round-robin manner
works well when the network bottleneck exists between the intermediary and the

> end client, as the intermediary would be buffering the responses and then be

Reviewer: ```suggestion
user agent, as the intermediary would be buffering the responses and then be
```

forwarding the chunks of those buffered responses based on the prioritization
scheme it implements. A sophisticated server MAY use a weighted round-robin
reflecting the urgencies expressed in the requests, so that less urgent
responses would receive less bandwidth in case the bottleneck exists between the
server and the intermediary.

### HTTP/1.x Back Ends

It is common for CDN infrastructure to support different HTTP versions on the
front end and back end. For instance, the client-facing edge might support
HTTP/2 and HTTP/3 while communication to back end servers is done using
HTTP/1.1. Unlike with connection coalescing, the CDN will "de-mux" requests into
discrete connections to the back end. As HTTP/1.1 and older do not provide a way
to concurrently transmit multiple responses, there is no immediate fairness
issue in protocol. However, back end servers MAY still use client headers for
request scheduling. Back end servers SHOULD only schedule based on client

> priority information where that information can be scoped to individual end

Reviewer: that's not clear to me
